tests/test_middleware_interface.py

```diff
@@ -362,6 +361,52 @@ def test_run_pipeline_empty_quantum_graph(self):
         with self.assertRaisesRegex(RuntimeError, "No data to process"):
             self.interface.run_pipeline(self.next_visit, {2})
 
+    def _assert_in_collection(self, butler, collection, dataset_type, data_id):
+        for dataset in butler.registry.queryDatasets(dataset_type, collections=collection, dataId=data_id):
+            return
+        self.fail(f"No datasets found matching {dataset_type}@{data_id} in {collection}.")
+
+    def _assert_not_in_collection(self, butler, collection, dataset_type, data_id):
+        for dataset in butler.registry.queryDatasets(dataset_type, collections=collection, dataId=data_id):
+            self.fail(f"{dataset} matches {dataset_type}@{data_id} in {collection}.")
+
+    def test_clean_unsafe_datasets(self):
+        """Test that _clean_unsafe_datasets removes only "conflicting" datasets.
```

Comment:
Worth noting here that "conflicting" datasets are anything without `detector` in their datasetType? I realize that's an implementation detail, but I had to look at it to understand what was "assert_in" and what was "assert_not_in" at the end.

Reply:
Well, the concept of a "conflicting" dataset is non-algorithmic: it's anything that would be produced with the same data ID (with or without different contents) for a different AP pipeline run. These just happen to be specific datasets I could think of that did/didn't meet those criteria.

Is there some other way to make the "in" vs. "not in" distinction more readable?

Reply:
Maybe a note to that effect in this docstring. Something like this? `The "conflicting" datasets in this test are specific examples that clearly meet the "can be produced with the same data ID in different AP runs" criteria; currently "contains detector in the data ID".`

Reply:
I went with:

> A conflicting dataset is one that can be produced with the same data ID in different AP runs. This test currently uses `calibrate_config` and `src_schema` as examples of conflicting datasets, and `src` and `calexp` as counterexamples (which should not be removed).